server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java

```diff
@@ -414,13 +426,72 @@ private TabletMetadata getTabletMetadata(KeyExtent extent) {
     }
   }
 
+  protected boolean insertScanReferences(TabletMetadata tm) {
+    final String serverAddress = clientAddress.toString();
+    final String clientAddress = TServerUtils.clientAddress.get();
+    final Set<ScanServerRefTabletFile> refsToAdd = new HashSet<>();
+    tm.getFiles().forEach(f -> {
+      AtomicInteger count = this.allReservedFiles.computeIfAbsent(f, x -> new AtomicInteger(0));
+      if (count.get() == 0) {
+        refsToAdd.add(new ScanServerRefTabletFile(f.getPathStr(), serverAddress, serverLockUUID,
+            clientAddress));
+      }
+      count.incrementAndGet();
+    });
+
+    if (refsToAdd.isEmpty()) {
+      return true;
+    }
+    try {
+      this.getContext().getAmple().putScanServerFileReferences(refsToAdd);
+    } catch (Exception e) {
+      LOG.error("Error inserting scan server references", e);
+      return false;
+    }
+    return true;
+  }
+
+  protected void deleteScanReferences(long sessionId, boolean batchScan) {
+    Tablet tablet = null;
+    if (!batchScan) {
+      SingleScanSession session = (SingleScanSession) sessionManager.getSession(sessionId);
+      tablet = session.getTabletResolver().getTablet(session.extent);
+    } else {
+      MultiScanSession session = (MultiScanSession) sessionManager.getSession(sessionId);
+      tablet = session.getTabletResolver().getTablet(session.threadPoolExtent);
+    }
+    deleteScanReferences(tablet);
+  }
+
+  protected void deleteScanReferences(Tablet tablet) {
+    final String serverAddress = clientAddress.toString();
+    final String clientAddress = TServerUtils.clientAddress.get();
+    final Set<StoredTabletFile> files = tablet.getDatafiles().keySet();
+    final Set<ScanServerRefTabletFile> filesToRemove = new HashSet<>();
+    files.forEach(f -> {
+      AtomicInteger count = this.allReservedFiles.get(f);
+      if (count != null && count.decrementAndGet() == 0) {
+        filesToRemove.add(new ScanServerRefTabletFile(f.getPathStr(), serverAddress, serverLockUUID,
+            clientAddress));
+      }
+    });
+    if (!filesToRemove.isEmpty()) {
+      this.getContext().getAmple().deleteScanServerFileReferences(filesToRemove);
+    }
+  }
+
   protected ScanInformation loadTablet(KeyExtent extent)
       throws IllegalArgumentException, IOException, AccumuloException {
 
     // Need to call ScanServerKeyExtent.toKeyExtent so that the cache key
     // is the unique per tablet, not unique per scan.
     TabletMetadata tabletMetadata = getTabletMetadata(extent);
 
+    while (!insertScanReferences(tabletMetadata)) {
+      tabletMetadataCache.invalidate(extent);
+      tabletMetadata = getTabletMetadata(extent);
+    }
+
```

Comment:
There is a an external race condition here.  We may insert scan entries for files a tablet has already dereferenced.  After we insert new scan server refs we need to verify the tablet is still referencing them.

Reply:
I thought the plan for eventually consistent scans was that you get the files at the time of the scan. Even if the master copy of the tablet has compacted and one of the files has been dereferenced from the tablet in the metadata table, the file will still exist in HDFS and this copy of the TabletMetadata still references it. What am I missing?

Reply:
I am thinking of the following case where between the time we read a tablets files from the metadata table and write the scan server refs for those files, the files are deleted.

 1. Scan server S1 gets a list of files [F1,F2,F3] for tablet T1
 2. Tablet T1 compacts files [F2,F3] to F4 and writes deletions candidate entries to metadata table for [F2, F3]
 3. GC deletes files [F2,F3]
 4. Scan server S1 writes refs for [F1,F2,F3] to the metadata table
 5. Scan server S1 starts a scan with files [F1,F2,F3] which fails because some files are missing.

Reply:
Sure, but that's a really small window and the garbage collector doesn't run that often

Reply:
It's a small window/probability for a single event, but if the situation happens frequently then the probability grows.  The more scan servers, tablets, compactions, and scans the higher the probability of hitting this.  I suspect on a large busy system it could happen regularly.